### PR TITLE
CP should close properly, even if IJ is not running

### PR DESCRIPTION
**Pull request by zindy**
I had a bug (Linux) when trying to exit CP, because IJ is not running.

Crude fix, but I don't understand how javabridge works. There may be a better way.

Error message without the fix:

```
Failed to stop the JVM
Traceback (most recent call last):
  File "/mnt/home/epzindy/projects/CellProfiler/cellprofiler/__main__.py", line 276, in stop_cellprofiler
    cp_stop_vm()
  File "/mnt/home/epzindy/projects/CellProfiler/cellprofiler/utilities/cpjvm.py", line 249, in cp_stop_vm
    ij1 = javabridge.JClassWrapper("ij.IJ")
  File "/mnt/home/epzindy/anaconda/lib/python2.7/site-packages/javabridge/wrappers.py", line 175, in __init__
    self.klass = J.get_class_wrapper(J.class_for_name(class_name), True)
  File "/mnt/home/epzindy/anaconda/lib/python2.7/site-packages/javabridge/jutil.py", line 1711, in class_for_name
    classname, True, ldr)
  File "/mnt/home/epzindy/anaconda/lib/python2.7/site-packages/javabridge/jutil.py", line 906, in static_call
    result = fn(*nice_args)
  File "/mnt/home/epzindy/anaconda/lib/python2.7/site-packages/javabridge/jutil.py", line 886, in fn
    raise JavaException(jexception)
JavaException: ij.IJ
```
... now moving onto *why* IJ is not running!

### CODE REVIEW

**Comment by LeeKamentsky** [comment]
Hi Egor,
I am guessing that you didn't do a `pip install prokaryote` to install the Java jar files. That's why it wouldn't be able to find the class files in order to close.

**Comment by zindy** [comment]
Thanks Lee,

The issue was that I didn't have a HTTPS_PROXY environment variable defined, only a HTTP_PROXY environment variable. I pulled prokaryote from github and checked how the setup.py file works.

`python setup.py fetch_prokaryote_jar` uses a *https* address rather than *http*.

I actually managed to run CellProfiler 4ae41bc on Linux. My pipeline (well, at least its first step) is working!

Thank you for your help. Not sure if this PR is all that useful then.

Kind regards,
Egor

**Comment by 0x00B1** [comment]
@zindy I think it’s still useful!

File: cellprofiler/utilities/cpjvm.py [hunk 1]
```diff
@@ -246,7 +246,15 @@ def cp_stop_vm(kill=True):
...
     '''
     from imagej.imagej2 import allow_quit, the_imagej_context
-    ij1 = javabridge.JClassWrapper("ij.IJ").getInstance()
+
+    #
+    # This may fail if ImageJ is not running
+    #
+    try:
+        ij1 = javabridge.JClassWrapper("ij.IJ").getInstance()
+    except:
```
Comment (0x00B1) [inline cellprofiler/utilities/cpjvm.py]: Specify the thrown exception

File: cellprofiler/utilities/cpjvm.py [hunk 2]
```diff
@@ -246,7 +246,15 @@ def cp_stop_vm(kill=True):
     whatever action is needed to stop AWT and the JVM.
     '''
     from imagej.imagej2 import allow_quit, the_imagej_context
-    ij1 = javabridge.JClassWrapper("ij.IJ").getInstance()
+
+    #
+    # This may fail if ImageJ is not running
```
Comment (0x00B1) [inline cellprofiler/utilities/cpjvm.py]: Unnecessary comments (especially when the exception is specified)

**Comment by 0x00B1** [comment]
@zindy Would it help if I made the changes I suggested?

**Comment by zindy** [comment]
Hi 0x00B1, I have removed the unnecessary comment. Merge if you wish.

Kind regards,
Egor

**Comment by 0x00B1** [comment]
@zindy The comment was unnecessary if you specified the exception:

```Python
try
    pass
except …
    pass
```

i.e. what exception are you trying to capture?

**Comment by zindy** [comment]
Sorry for misunderstanding you.

I have amended the code so that an error is logged IF the specific exception javabridge.JavaException is raised when trying to get the ij.IJ instance.

In which case, the logged message should be:
No available instance: ij.IJ

The intent of the code should now be evident, so no comments needed. Let me know if this is OK.

Kind regards,
Egor

**Comment by 0x00B1** [comment]
Perfect, thanks!

**Pull request merged**